src/infra/core/src/query/mod.rs

```diff
@@ -74,7 +76,14 @@ pub(crate) struct KamuSchema {
 }
 
 struct KamuSchemaImpl {
-    session_context: SessionContext,
+    session_config: SessionConfig,
+    table_options: TableOptions,
+    // To avoid creating an Arc cyclic reference[1], with SessionContext, we are forced to use
+    // weak references instead.
+    //
+    // [1] SessionContext -> KamuCatalog(CatalogProvider) -> KamuSchema -> KamuSchemaImpl
+    //     -> SchemaCache -> KamuTable -> SessionContext
+    session_state: Weak<RwLock<SessionState>>,
```

Comment:
Amazing job at figuring this out!
I didn't think about circular references and imagine this was super hard to find.

I think the actual solution can be simplifed to possibly avoid shared state completely. I will think about it more and suggest some options.

Reply:
> I think the actual solution can be simplifed to possibly avoid shared state completely. I will think about it more and suggest some options.

please note that a `DataFrame` cannot exist without a `SessionContext`. Or in other words, `SessionContext` can (should) exist longer than `DataFrame`